### DISCUSSION TITLE
How to insert a json value in a supabase function?

### DISCUSSION THREAD

**Discussion by KevTale** [supabase/supabase#3524] [Questions] [answered]
Hi,

I've got this trigger with an associated function:

```

create function public.handle_new_submission()
returns trigger as $$
begin
  insert into public.submissions (id, contest_occurrence, winner)
  values (new.id, 1, ???);
  return new;
end;
$$ language plpgsql security definer;


create trigger on_submission_inserted
  after insert on public.submissions
  for each row execute procedure public.handle_new_submission();
```

I want `winner` (currently the " ??? ") value to looks like this:

```
{
  is_winner: false,
  contest_occurrence: 1,
 }
```

But I have no idea how to write this and not sure how to effectively google it...

Can someone help?

Thanks!

**Comment by silentworks** [comment]
It would be best to state what you are trying to accomplish with this, it is hard to direct you without understanding the context or use case.

**Reply by KevTale** [reply]
I'll give you the full context @silentworks , however after doing some digging it may seem like my design structure is not the best one.

Here's the full context:

I'm building a webapp where users can submit painting for a contest. Then, each one of them can upvote they preferred ones and by the end of each week the winner (most upvoted one) get a reward.

**To participate (submit and vote), users must be authenticated. Users can only upvote once per submission.**

Here is my structure (it differs a bit from my initial post):

### `painting_submissions` table:

| id (int8) 	|            artist (jsonb)           	| painting_ref (text) 	| inserted_at 	| updated_at 	|
|:---------:	|:-----------------------------------:	|---------------------	|-------------	|------------	|
|     1     	| {username: 'Picasso', name: 'John'} 	| 1f3d4Mf6e7        	| ...         	| ...        	|

This table has `SELECT` access for everyone, `insert` for `authenticated` users.

`painting_ref ` is the unique identifier corresponding to the piece of art itself which is stored in a separate storage platform.

Here is how I make the request in JS:

```
 const { data, error } = await supabase.from("painting_submissions").insert(
{
  artist: { username: user?.username, name: user?.name },
  painting_ref
}, { returning: "minimal" });
```

### `votes` table:

To keep things secured, I figured I would make the votes counter in a separated table which blocks users to directly `insert` malicious submissions or `updates` existing ones.

| id (int8) 	| number_of_votes (int, default to 0) 	| painting_ref (text) 	| already_votes_uid (varchar[], default to empty array) 	| inserted_at 	| updated_at 	|
|:---------:	|:----------------------:	|---------------------	|-------------------------------	|-------------	|------------	|
|     1     	|           12           	| 1f3d4Mf6e7          	| ["123", "456", "789"]         	| ...         	| ...        	|

This table has RLS on and read access to everyone but no INSERT or UPDATE.

My idea is to make a trigger that listen when there is an `insert` in `painting_submissions` then insert a new row in `votes`.

```
create function public.insert_new_vote_row()
returns trigger as $$
begin
  insert into public.votes (id, painting_ref) <-- how can I grab the painting_ref from my initial insert call?
  values (new.id, painting_ref);
  return new;
end;
$$ language plpgsql security definer;

create trigger on_painting_submission_inserted_insert_new_vote_row
  after insert on public.painting_submissions
  for each row execute procedure public.insert_new_vote_row();
```

Once done, the `votes` table would have a new row with the defaulted values for `number_of_votes` and `already_voted_uid`. **Admitting this pattern is correct, I have a first blocking problem here, I'm not sure how to grab `painting_ref` on my `insert_new_vote_row` function, knowing that this value is being send by the first `supabase.from("painting_submissions").insert(...)` call.**

### Voting system

My idea is to trigger an `upvote` function using `rpc`.

```
create function upvote (painting_ref_to_upvote text) 
returns void as
$$
  update votes 
  set number_of_votes = number_of_votes + 1
  where painting_ref = painting_ref_to_upvote;
$$ 
language sql volatile;
```

```
await supabase.rpc("upvote", {
 painting_ref_to_upvote: painting_ref,
});
```

...which works well.

**On top of that, I should somehow populate the `already_voted_uid` array of the `votes` table when the voting occurs.** I'm not sure how to proceed. I could somehow (in pseudocode) `set already_voted_uid = already_votes_uid.push(new_uid)` or create a new `already_votes_uid` which has `painting_ref` and `uid` columns.

Either way, user should be able to vote only if he has not voted yet. So I can make a check in frontend, grabbing the `already_voted_uid` array and checking if his `uid` is included in it. But for security sake, I should also make that check sql side, shouldn't I? Probably in the `upvote` function. But I'm not sure how.

I think that's pretty much it... I hope I was clear enough!
Thank you for your time @silentworks , I know this is a long post 😅

**Reply by fergusmeiklejohn** [reply]
I would solve this differently and I'd be interested to get feedback on this solution..
Every user would be authenticated and we'd have a public.users table which referenced auth.users.
public.users would have user_name, name, user_id etc
then there would be a table for artwork with artwork_id, name, user_id foreign key on public.users
Then a voting table with artwork_id FK on public.artwork.artwork_id and user_id FK on public.users.user_id, then we can make a primary key from these two columns so no user can ever twice for the same artwork, and RLS + policy so only an authenticated user can write to a row with their user_id
To see which artwork got the most votes we group by artwork_id and count the rows

**Comment by burggraf** [comment, marked as answer]
Look into using `JSON` (or better yet `JSONB`) with PostgreSQL.
https://supabase.io/docs/guides/database/json

For your example:

```sql
insert into public.submissions (id, contest_occurrence, winner) 
values (new.id, 1, ???);
```

you could probably just do:
```sql
insert into public.submissions (id, contest_occurrence, winner)
  values (new.id, 1, '{is_winner: false, contest_occurrence: 1}');
```
